tests/unit/test_endpoints_v2.py

```diff
@@ -0,0 +1,634 @@
+                    ]
+                },
+            ),
+        ),
+    ],
+)
+def test_endpoint_resolution_with_is_valid_host_label(
+    partitions, ruleset_dict, input_parameters, expected_endpoint
+):
+    endpoint_provider = EndpointProvider(ruleset_dict, partitions)
+    endpoint = endpoint_provider.resolve_endpoint(**input_parameters)
+    assert endpoint.url == expected_endpoint.url
+    assert endpoint.properties == expected_endpoint.properties
+    assert endpoint.headers == expected_endpoint.headers
+
+
+def test_endpoint_resolution_invalid_host_label(partitions, ruleset_dict):
+    params = {
+        "Region": "us-east-1",
+        "Bucket": "arn:aws:s3:us-west-2:123456789012:accesspoint:myendpoint",
+        "HostLabel": "foo://this.is.not:.a.\va!lid.host.label",
+    }
+    endpoint_provider = EndpointProvider(ruleset_dict, partitions)
+    with pytest.raises(EndpointResolutionError) as exc_info:
+        endpoint_provider.resolve_endpoint(**params)
+    assert str(exc_info.value) == "Default error message"
+
+
+@pytest.mark.parametrize(
+    "label,allow_subdomains,expected_output",
+    [
+        (None, False, False),
+        (None, True, False),
+        ("test.me.please.com", False, False),
+        ("test.me.please.com", True, True),
+        ("s3.us-east-1.amazonaws.com", True, True),
+        ("s3.us-east-1.amazonaws.com", False, False),
+        ("amazon.com", False, False),
+        ("amazon.com", True, True),
+        ("foo", False, True),
+        ("foo", True, True),
+    ],
+)
+def test_is_valid_host_label(
+    standard_library, label, allow_subdomains, expected_output
+):
+    assert (
+        standard_library._is_valid_host_label(label, allow_subdomains)
+        == expected_output
+    )
+
+
+def test_partition_none_return_none(standard_library):
+    assert standard_library._aws_partition(None) is None
+
+
+@pytest.mark.parametrize("input_arn", [None, "arn:aws:this-is-not-an-arn:foo"])
+def test_parse_arn_none_return_none(standard_library, input_arn):
+    assert standard_library._aws_parse_arn(input_arn) is None
+
+
+@pytest.mark.parametrize(
+    "uri,expected_output",
+    [
+        (None, None),
+        (
+            ":/?#[]@!$&'()*+,;=%.",
+            "%3A%2F%3F%23%5B%5D%40%21%24%26%27%28%29%2A%2B%2C%3B%3D%25.",
+        ),
+        ("?foo=bar&baz=qux", "%3Ffoo%3Dbar%26baz%3Dqux"),
+    ],
+)
+def test_uri_encode(standard_library, uri, expected_output):
+    assert standard_library._uri_encode(uri) == expected_output
+
+
+def test_parse_url_none_return_none(standard_library):
+    assert standard_library._parse_url(None) is None
+
+
+def test_string_equals_wrong_type_raises(standard_library):
+    with pytest.raises(RuleSetError) as exc_info:
+        standard_library._string_equals(1, 2)
+    assert str(exc_info.value) == (
+        "Something went wrong during endpoint resolution: Both values must be strings.\n"
+        "value1: 1\n"
+        "value2: 2"
+    )
```

Comment:
I think it would. This would cause rule set evaluation to break, not resolve to an error rule. I'll update to be similar to the other suggestion you made.